File: vms/proxy.go
package main

import (
	"bufio"
	"crypto/rand"
	"encoding/hex"
	"flag"
	"fmt"
	"io"
	"log"
	"net"
	"net/http"
	"net/url"
	"strings"
	"sync/atomic"
	"time"
)

var (
	listenAddr    = flag.String("l", "127.0.0.1:9902", "Address to listen on")
	upstreamProxy = flag.String("p", "http://127.0.0.1:8119", "Upstream proxy URL")
	whitelist     = flag.String("w", "", "Comma-separated list of domains to whitelist")
)

// Whitelist for domain checking
type Whitelist struct {
	exactMatches  map[string]struct{}
	suffixMatches []string
}

// CounterConn wraps a net.Conn to count bytes read and written.
type CounterConn struct {
	net.Conn
	readBytes    atomic.Int64
	writtenBytes atomic.Int64
}

// NewCounterConn creates a new CounterConn.
func NewCounterConn(conn net.Conn) *CounterConn {
	return &CounterConn{Conn: conn}
}

// Read wraps the underlying Read method to count bytes.
func (c *CounterConn) Read(b []byte) (int, error) {
	n, err := c.Conn.Read(b)
	if n > 0 {
		c.readBytes.Add(int64(n))
	}
	return n, err
}

// Write wraps the underlying Write method to count bytes.
func (c *CounterConn) Write(b []byte) (int, error) {
	n, err := c.Conn.Write(b)
	if n > 0 {
		c.writtenBytes.Add(int64(n))
	}
	return n, err
}

// GetReadBytes returns the total bytes read.
func (c *CounterConn) GetReadBytes() int64 {
	return c.readBytes.Load()
}

// GetWrittenBytes returns the total bytes written.
func (c *CounterConn) GetWrittenBytes() int64 {
	return c.writtenBytes.Load()
}

func parseWhitelist(list string) *Whitelist {
	wl := &Whitelist{
		exactMatches:  make(map[string]struct{}),
		suffixMatches: []string{},
	}

	domains := strings.Split(list, ",")
	for _, domain := range domains {
		domain = strings.TrimSpace(strings.ToLower(domain))
		if domain == "" {
			continue
		}

		if strings.HasPrefix(domain, "*.") {
			suffix := domain[2:] // Remove "*." prefix
			wl.suffixMatches = append(wl.suffixMatches, suffix)
		} else {
			wl.exactMatches[domain] = struct{}{}
		}
	}

	return wl
}

func (wl *Whitelist) isAllowed(domain string) bool {
	domain = strings.TrimSpace(strings.ToLower(domain))

	// Check exact match
	if _, ok := wl.exactMatches[domain]; ok {
		return true
	}

	// Check suffix match
	for _, suffix := range wl.suffixMatches {
		if domain == suffix || strings.HasSuffix(domain, "."+suffix) {
			return true
		}
	}

	return false
}

func main() {
	flag.Parse()

	// Parse the whitelist
	wl := parseWhitelist(*whitelist)

	// Parse the upstream proxy URL
	upstreamURL, err := url.Parse(*upstreamProxy)
	if err != nil {
		log.Fatalf("Failed to parse upstream proxy URL: %v", err)
	}

	// Create the proxy server
	server := &ProxyServer{
		whitelist:   wl,
		upstreamURL: upstreamURL,
	}

	// Start listening
	log.Printf("HTTP proxy server listening on %s", *listenAddr)
	log.Fatal(http.ListenAndServe(*listenAddr, server))
}

// ProxyServer handles both CONNECT tunneling and standard HTTP proxying
type ProxyServer struct {
	whitelist   *Whitelist
	upstreamURL *url.URL
}

func (p *ProxyServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	// Extract the domain from the request
	var domain string
	if r.Method == http.MethodConnect {
		host, _, err := net.SplitHostPort(r.Host)
		if err != nil {
			domain = r.Host
		} else {
			domain = host
		}
	} else {
		if r.URL.Host != "" {
			host, _, err := net.SplitHostPort(r.URL.Host)
			if err != nil {
				domain = r.URL.Host
			} else {
				domain = host
			}
		} else {
			host, _, err := net.SplitHostPort(r.Host)
			if err != nil {
				domain = r.Host
			} else {
				domain = host
			}
		}
	}

	// Check if the domain is in the whitelist
	if !p.whitelist.isAllowed(domain) {
		log.Printf("Domain %s is not in the whitelist", domain)
		http.Error(w, "Domain not allowed", http.StatusForbidden)
		return
	}

	// Handle the request based on the method
	if r.Method == http.MethodConnect {
		p.handleConnect(w, r)
	} else {
		p.handleStandardRequest(w, r)
	}
}

func (p *ProxyServer) handleConnect(w http.ResponseWriter, r *http.Request) {
	// Generate a unique trace ID using the standard library's crypto/rand.
	var traceID string
	randomBytes := make([]byte, 16) // 128 bits of randomness
	if _, err := rand.Read(randomBytes); err != nil {
		// Fallback to a timestamp-based ID if crypto/rand fails.
		log.Printf("WARNING: Could not generate random bytes for traceID: %v. Falling back to timestamp.", err)
		traceID = fmt.Sprintf("fallback-%d", time.Now().UnixNano())
	} else {
		traceID = hex.EncodeToString(randomBytes)
	}

	startTime := time.Now()
	log.Printf("[%s] New TCP connection for %s from %s", traceID, r.Host, r.RemoteAddr)

	// Connect to the upstream proxy
	upstreamConn, err := net.Dial("tcp", p.upstreamURL.Host)
	if err != nil {
		log.Printf("[%s] Failed to connect to upstream proxy: %v", traceID, err)
		http.Error(w, "Failed to connect to upstream proxy", http.StatusBadGateway)
		return
	}

	// Send a CONNECT request to the upstream proxy
	upstreamReq := &http.Request{
		Method: http.MethodConnect,
		URL:    &url.URL{Host: r.Host},
		Host:   r.Host,
		Header: make(http.Header),
	}

	if err := upstreamReq.Write(upstreamConn); err != nil {
		log.Printf("[%s] Failed to write CONNECT request to upstream proxy: %v", traceID, err)
		http.Error(w, "Failed to establish tunnel", http.StatusBadGateway)
		upstreamConn.Close()
		return
	}

	// Read the response from the upstream proxy
	upstreamReader := bufio.NewReader(upstreamConn)
	upstreamResp, err := http.ReadResponse(upstreamReader, upstreamReq)
	if err != nil {
		log.Printf("[%s] Failed to read response from upstream proxy: %v", traceID, err)
		http.Error(w, "Failed to establish tunnel", http.StatusBadGateway)
		upstreamConn.Close()
		return
	}
	defer upstreamResp.Body.Close()

	// Check if the tunnel was established successfully
	if upstreamResp.StatusCode != http.StatusOK {
		log.Printf("[%s] Upstream proxy failed to establish tunnel: %s", traceID, upstreamResp.Status)
		http.Error(w, "Failed to establish tunnel", upstreamResp.StatusCode)
		upstreamConn.Close()
		return
	}

	// Hijack the client connection
	hj, ok := w.(http.Hijacker)
	if !ok {
		log.Printf("[%s] Client connection does not support hijacking", traceID)
		http.Error(w, "Tunnel cannot be established", http.StatusInternalServerError)
		upstreamConn.Close()
		return
	}
	clientConn, _, err := hj.Hijack()
	if err != nil {
		log.Printf("[%s] Failed to hijack client connection: %v", traceID, err)
		http.Error(w, "Tunnel cannot be established", http.StatusInternalServerError)
		upstreamConn.Close()
		return
	}

	// Send 200 OK response to the client to indicate tunnel established
	if _, err := clientConn.Write([]byte("HTTP/1.1 200 Connection established\r\n\r\n")); err != nil {
		log.Printf("[%s] Failed to send 200 OK to client: %v", traceID, err)
		clientConn.Close()
		upstreamConn.Close()
		return
	}

	// Wrap connections to count traffic
	countingClientConn := NewCounterConn(clientConn)
	countingUpstreamConn := NewCounterConn(upstreamConn)

	// Set up bidirectional data transfer
	done := make(chan struct{})

	go func() {
		io.Copy(countingUpstreamConn, countingClientConn)
		done <- struct{}{}
	}()

	go func() {
		io.Copy(countingClientConn, countingUpstreamConn)
		done <- struct{}{}
	}()

	// Wait for either direction to finish
	<-done

	// Close connections
	clientConn.Close()
	upstreamConn.Close()

	// Log the end of the connection and traffic stats
	duration := time.Since(startTime)
	uploadedBytes := countingClientConn.GetReadBytes()
	downloadedBytes := countingClientConn.GetWrittenBytes()

	log.Printf(
		"[%s] Connection closed. Duration: %s, Upload: %s, Download: %s",
		traceID,
		duration.Round(time.Millisecond),
		formatBytes(uploadedBytes),
		formatBytes(downloadedBytes),
	)
}

// formatBytes is a helper to format byte counts into KB, MB, etc.
func formatBytes(b int64) string {
	const unit = 1024
	if b < unit {
		return fmt.Sprintf("%d B", b)
	}
	div, exp := int64(unit), 0
	for n := b / unit; n >= unit; n /= unit {
		div *= unit
		exp++
	}
	return fmt.Sprintf("%.2f %ciB", float64(b)/float64(div), "KMGTPE"[exp])
}

func (p *ProxyServer) handleStandardRequest(w http.ResponseWriter, r *http.Request) {
	log.Printf("Handling standard HTTP request for %s from %s", r.Host, r.RemoteAddr)
	transport := &http.Transport{
		Proxy: http.ProxyURL(p.upstreamURL),
	}
	client := &http.Client{
		Transport: transport,
		CheckRedirect: func(req *http.Request, via []*http.Request) error {
			return http.ErrUseLastResponse
		},
	}

	outReq := new(http.Request)
	*outReq = *r
	outReq.RequestURI = ""

	if !r.URL.IsAbs() {
		outReq.URL = &url.URL{
			Scheme:   "http",
			Host:     r.Host,
			Path:     r.URL.Path,
			RawQuery: r.URL.RawQuery,
		}
		if r.TLS != nil {
			outReq.URL.Scheme = "https"
		}
	}

	resp, err := client.Do(outReq)
	if err != nil {
		log.Printf("Failed to execute request: %v", err)
		http.Error(w, "Failed to reach destination", http.StatusBadGateway)
		return
	}
	defer resp.Body.Close()

	for key, values := range resp.Header {
		for _, value := range values {
			w.Header().Add(key, value)
		}
	}
	w.WriteHeader(resp.StatusCode)
	io.Copy(w, resp.Body)
}

File: vms/proxy.py
import argparse
import logging
import os
import select
import shutil
import socket
import time
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from urllib.parse import urlparse
from urllib import request as urllib_request
from urllib import error as urllib_error

# 配置日志
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
log = logging.getLogger(__name__)

class Whitelist:
    """处理域名白名单的类"""
    def __init__(self, list_str: str):
        self.exact_matches = set()
        self.suffix_matches = []

        domains = [d.strip().lower() for d in list_str.split(',') if d.strip()]
        for domain in domains:
            if domain.startswith('*.'):
                self.suffix_matches.append(domain[2:])
            else:
                self.exact_matches.add(domain)

    def is_allowed(self, domain: str) -> bool:
        """检查域名是否在白名单内"""
        domain = domain.strip().lower()

        # 1. 检查精确匹配
        if domain in self.exact_matches:
            return True

        # 2. 检查后缀匹配
        for suffix in self.suffix_matches:
            if domain == suffix or domain.endswith(f".{suffix}"):
                return True

        return False

class CounterSocketWrapper:
    """一个包装 socket 的类，用于统计读写的字节数"""
    def __init__(self, sock: socket.socket):
        self._socket = sock
        self.read_bytes = 0
        self.written_bytes = 0

    def recv(self, bufsize: int) -> bytes:
        data = self._socket.recv(bufsize)
        self.read_bytes += len(data)
        return data

    def sendall(self, data: bytes):
        self._socket.sendall(data)
        self.written_bytes += len(data)

    # 将其他 socket 方法委托给内部的 socket 对象
    def __getattr__(self, name):
        return getattr(self._socket, name)

def format_bytes(b: int) -> str:
    """将字节数格式化为可读的字符串 (KB, MB, etc.)"""
    if b < 1024:
        return f"{b} B"
    unit = 1024
    exp = 0
    div = unit
    n = b / unit
    while n >= unit:
        div *= unit
        exp += 1
        n /= unit
    return f"{b/div:.2f} {'KMGTPE'[exp]}iB"

class ProxyRequestHandler(BaseHTTPRequestHandler):
    """代理请求处理器"""
    # 这些类属性将在 main 函数中被设置
    whitelist: Whitelist = None
    upstream_url: 'urllib.parse.ParseResult' = None

    # 覆盖默认的 server header
    server_version = "PythonProxy/0.1"
    sys_version = ""

    def log_message(self, format: str, *args):
        # 禁用默认的 BaseHTTPRequestHandler 日志，我们使用自己的 logging 模块
        pass

    def do_CONNECT(self):
        """处理 HTTPS 的 CONNECT 请求"""
        trace_id = os.urandom(16).hex()
        start_time = time.monotonic()

        try:
            host, port_str = self.path.split(':', 1)
            port = int(port_str)
        except ValueError:
            self.send_error(400, "Bad CONNECT request")
            return

        log.info(f"[{trace_id}] New TCP connection for {self.path} from {self.client_address}")

        if not self.whitelist.is_allowed(host):
            log.warning(f"[{trace_id}] Domain {host} is not in the whitelist")
            self.send_error(403, "Domain not allowed")
            return

        # 1. 连接到上游代理
        try:
            upstream_socket = socket.create_connection(
                (self.upstream_url.hostname, self.upstream_url.port or 80)
            )
        except Exception as e:
            log.error(f"[{trace_id}] Failed to connect to upstream proxy: {e}")
            self.send_error(502, "Upstream connection failed")
            return

        try:
            # 2. 向上游代理发送 CONNECT 请求以建立隧道
            connect_req = f"CONNECT {self.path} HTTP/1.1\r\nHost: {self.path}\r\n\r\n"
            upstream_socket.sendall(connect_req.encode('utf-8'))

            # 3. 读取上游代理的响应
            upstream_file = upstream_socket.makefile('rb', 0)
            status_line = upstream_file.readline().decode('utf-8').strip()
            version, code, message = status_line.split(' ', 2)
            if int(code) != 200:
                log.error(f"[{trace_id}] Upstream proxy failed: {status_line}")
                self.send_error(int(code), message)
                return

            # 读取并忽略剩余的 header
            while True:
                if upstream_file.readline() in (b'\r\n', b'\n', b''):
                    break

            # 4. 向上游隧道建立成功，向客户端发送成功响应
            self.send_response(200, 'Connection established')
            self.end_headers()
        except Exception as e:
            log.error(f"[{trace_id}] Failed to establish tunnel with upstream: {e}")
            self.send_error(502, "Tunnel establishment failed")
            return
        finally:
            if 'upstream_file' in locals():
                # makefile() 会缓冲数据, 关闭它确保所有数据都被处理
                upstream_file.close()

        # 5. 在客户端和上游代理之间双向转发数据
        client_conn = self.connection
        client_counter = CounterSocketWrapper(client_conn)
        upstream_counter = CounterSocketWrapper(upstream_socket)

        try:
            sockets = [client_counter, upstream_counter]
            while sockets:
                readable, _, exceptional = select.select(sockets, [], sockets, 60)
                if exceptional:
                    break

                for sock in readable:
                    other_sock = upstream_counter if sock is client_counter else client_counter
                    data = sock.recv(8192)
                    if not data:
                        # 连接已关闭，停止转发
                        sockets.clear()
                        break
                    other_sock.sendall(data)
        except Exception as e:
            log.debug(f"[{trace_id}] Data transfer error: {e}")
        finally:
            # 6. 清理连接并记录统计信息
            duration = time.monotonic() - start_time
            uploaded = client_counter.read_bytes
            downloaded = client_counter.written_bytes
            log.info(
                f"[{trace_id}] Connection closed. Duration: {duration:.3f}s, "
                f"Upload: {format_bytes(uploaded)}, Download: {format_bytes(downloaded)}"
            )
            client_conn.close()
            upstream_socket.close()

    def handle_standard_request(self):
        """处理标准的 HTTP 请求 (GET, POST, etc.)"""
        # 1. 解析 URL 并检查白名单
        parsed_path = urlparse(self.path)
        host = parsed_path.hostname or self.headers.get('Host', '').split(':', 1)[0]

        if not self.whitelist.is_allowed(host):
            log.warning(f"Domain {host} is not in the whitelist")
            self.send_error(403, "Domain not allowed")
            return

        log.info(f"Handling standard HTTP request for {host} from {self.client_address}")

        # 2. 配置 urllib 使用上游代理
        proxy_handler = urllib_request.ProxyHandler({
            'http': self.upstream_url.geturl(),
            'https': self.upstream_url.geturl(),
        })

        # 禁用重定向，以匹配 Go 程序的行为
        class NoRedirectHandler(urllib_request.HTTPRedirectHandler):
            def redirect_request(self, req, fp, code, msg, headers, newurl):
                return None
            http_error_301 = http_error_302 = http_error_303 = http_error_307 = redirect_request

        opener = urllib_request.build_opener(proxy_handler, NoRedirectHandler)

        # 3. 构建发往上游的请求
        body = None
        if 'Content-Length' in self.headers:
            content_len = int(self.headers['Content-Length'])
            body = self.rfile.read(content_len)

        # 客户端发来的通常是相对路径，需要构造成完整 URL
        url_to_fetch = self.path
        if not url_to_fetch.startswith(('http://', 'https://')):
            url_to_fetch = f"http://{host}{self.path}"

        out_req = urllib_request.Request(
            url_to_fetch,
            data=body,
            headers=dict(self.headers),
            method=self.command
        )

        try:
            # 4. 发送请求并获取响应
            with opener.open(out_req) as resp:
                # 5. 将响应发回给客户端
                self.send_response(resp.code, resp.reason)
                for key, value in resp.getheaders():
                    # 过滤掉逐跳(hop-by-hop)的头信息
                    if key.lower() not in ('transfer-encoding', 'connection', 'content-encoding'):
                        self.send_header(key, value)
                self.end_headers()
                shutil.copyfileobj(resp, self.wfile)

        except urllib_error.HTTPError as e:
            # 处理来自目标服务器的错误响应 (如 404, 500)
            self.send_response(e.code, e.reason)
            for key, value in e.headers.items():
                 if key.lower() not in ('transfer-encoding', 'connection', 'content-encoding'):
                    self.send_header(key, value)
            self.end_headers()
            self.wfile.write(e.read())
        except Exception as e:
            log.error(f"Failed to execute standard request: {e}")
            self.send_error(502, "Proxy request failed")

    # 将所有常见的 HTTP 方法都指向同一个处理器
    def do_GET(self): self.handle_standard_request()
    def do_POST(self): self.handle_standard_request()
    def do_HEAD(self): self.handle_standard_request()
    def do_PUT(self): self.handle_standard_request()
    def do_DELETE(self): self.handle_standard_request()
    def do_OPTIONS(self): self.handle_standard_request()
    def do_PATCH(self): self.handle_standard_request()

def main():
    """主函数，用于解析参数和启动服务器"""
    parser = argparse.ArgumentParser(description="A simple Python HTTP/HTTPS proxy server.")
    parser.add_argument('-l', dest='listen_addr', default='127.0.0.1:9902', help='Address to listen on (default: 127.0.0.1:9902)')
    parser.add_argument('-p', dest='upstream_proxy', default='http://127.0.0.1:8119', help='Upstream proxy URL (default: http://127.0.0.1:8119)')
    parser.add_argument('-w', dest='whitelist', default='', help='Comma-separated list of whitelisted domains (e.g., "example.com,*.google.com")')
    args = parser.parse_args()

    # 将解析后的参数设置到请求处理器类上
    ProxyRequestHandler.whitelist = Whitelist(args.whitelist)
    ProxyRequestHandler.upstream_url = urlparse(args.upstream_proxy)

    if not ProxyRequestHandler.upstream_url.scheme or not ProxyRequestHandler.upstream_url.hostname:
        log.fatal(f"Invalid upstream proxy URL: {args.upstream_proxy}")
        return

    listen_host, listen_port_str = args.listen_addr.split(':')
    server_address = (listen_host, int(listen_port_str))

    httpd = ThreadingHTTPServer(server_address, ProxyRequestHandler)

    log.info(f"HTTP proxy server listening on {args.listen_addr}")
    log.info(f"Upstream proxy configured at {args.upstream_proxy}")
    if args.whitelist:
        log.info(f"Whitelist is active for domains: {args.whitelist}")
    else:
        log.warning("Whitelist is empty. All domains will be rejected unless you add domains with -w.")


    try:
        httpd.serve_forever()
    except KeyboardInterrupt:
        log.info("Shutting down server...")
    finally:
        httpd.server_close()

if __name__ == '__main__':
    main()
